// nest-telegraf/build/TelegrafModule.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var TelegrafModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
const telegraf_1 = require("telegraf");
const common_1 = require("@nestjs/common");
const core_1 = require("@nestjs/core");
const BotController_1 = require("./decorator/BotController");
let TelegrafModule = TelegrafModule_1 = class TelegrafModule {
    static forRootAsync({ useFactory, inject, imports }) {
        const providers = [
            {
                provide: telegraf_1.default,
                useFactory: async (ref, ...inject) => {
                    return Promise
                        .resolve()
                        .then(() => useFactory(...inject))
                        .then(({ token, options }) => {
                        return this.init(ref, token, options);
                    });
                },
                inject: [core_1.ModuleRef, ...(inject || [])],
            },
        ];
        return {
            imports,
            module: TelegrafModule_1,
            providers: providers,
            exports: providers,
        };
    }
    static forRoot(token, options) {
        const providers = [
            {
                provide: telegraf_1.default,
                useFactory: async (ref) => {
                    return this.init(ref, token, options);
                },
                inject: [core_1.ModuleRef],
            },
        ];
        return {
            module: TelegrafModule_1,
            providers: providers,
            exports: providers,
        };
    }
    static async init(ref, token, options) {
        const telegraf = new telegraf_1.default(token, options);
        BotController_1.handlers
            .filter((handler) => {
            return BotController_1.controllers.find((controller) => {
                return controller.class === handler.class;
            });
        })
            .forEach((handler) => {
            const object = ref.get(handler.class, { strict: false });
            if (!object) {
                return;
            }
            const func = Object.getOwnPropertyDescriptor(Object.getPrototypeOf(object), handler.method);
            if (!func) {
                return;
            }
            if (handler.type === 'hears') {
                const triggers = handler.property[0];
                return telegraf.hears(triggers, (ctx, next) => {
                    const ctxProperty = BotController_1.properties.find((property) => {
                        return property.target === Object.getPrototypeOf(object) && property.propertyKey === func.value.name && property.type === 'context';
                    });
                    const nextProperty = BotController_1.properties.find((property) => {
                        return property.target === Object.getPrototypeOf(object) && property.propertyKey === func.value.name && property.type === 'next';
                    });
                    const args = [];
                    if (ctxProperty) {
                        args[ctxProperty.parameterIndex] = ctx;
                    }
                    if (nextProperty) {
                        args[nextProperty.parameterIndex] = next;
                    }
                    const object2 = ref.get(handler.class, { strict: false });
                    return func.value.apply(object2, args);
                });
            }
            return;
        });
        return telegraf;
    }
};
TelegrafModule = TelegrafModule_1 = __decorate([
    common_1.Module({})
], TelegrafModule);
exports.default = TelegrafModule;
